### PR TITLE
#77 Create Form in Onboarding Page

### LINKED ISSUE #77: Create The Form That Goes Into The New OnBoarding Page And Handle Submission
**Related Issues**
- Project: #23
- Depends on: 
  - #76

**Description**
- Create a form, in the new OnBoarding page, with each of the fields from the user profile. This will include the following: `bio`, `twitter_username`, `linkedin_url`, `github_username`, `website`
 - `bio` should have (string, max 1000 chars), other fields should have (string, max 200 chars)

- Create a submit button on the page that updates the user profile fields with what they have completed, using the `PATCH /api/users/:id` endpoint.
- Redirects users to the dev-directory list page after submission

**Acceptance Criteria**
User can complete and submit the form.
The user model is updated.
The user is successfully redirected.

Includes end-to-end tests.

### PR DESCRIPTION
**Pull request by jmoldyvan**
Closes #77 

# Description

This PR creates a form in the onboarding page and submits user information to be included in the database/profile of the user.

## Testing

completed form and checked the profile url to see if it has appeared and is correct

E2E cypress testing complete and successful 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] Changes have new/updated automated tests, if applicable
- [X] Changes have new/updated docs, if applicable
- [X] I have performed a self-review of my own code
- [X] I have added comments on any new, hard-to-understand code
- [X] Changes have been manually tested
- [X] Changes generate no new errors or warnings

### CODE REVIEW

**Comment by Caleb-Cohen** [comment]
WHERE TESTING

**Comment by jmoldyvan** [comment]
> WHERE TESTING

HOL UP NEED TO MAKE IT BRB

**Comment by jmoldyvan** [comment]
alright got the testing

File: pages/onboarding/index.tsx
Comment (windyScripts) [inline pages/onboarding/index.tsx:1]: nit: Change UseridData to UserIdData, fits convention and is (slightly) more readable.

**Review by windyScripts — APPROVED** [review]
Beyond the nit, LGTM

**Comment by Caleb-Cohen** [comment]
Will add more comments later, but does the test confirm that the user profile is updated?

File: pages/onboarding/index.tsx
Comment (devonzara) [reply pages/onboarding/index.tsx:1]: I'd probably opt for `Props` or something instead, since it represents more than just the User that is held within.

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:33]: Initializing all these to be empty string will overwrite any existing data on the onboarding page - the default values here should be taken from the existing user data.

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:50]: 2xx status codes are technically okay, 4xx and above are errors.  Additionally, the `axios.patch` call above will error if the request failed, so this error is unreachable.

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:52]: i think we want to redirect them to `/directory` instead? rather than back to the home page

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:67]: i think it would also be best to put the skip button at the end of the form, rather than at the top - you'll see less user engagement if you present them with the option to skip right away

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:81]: all these fields are required, but i'm not convinced they should be? e.g. why should a user be _required_ to enter a twitter username?

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:83]: placeholders i think would be helpful here, e.g. so users don't also include the `@`

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:122]: this button should be disabled if there is an in-flight request - it will prevent them from double-clicking it accidentally.

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:26]: per your `getInitialProps`, `user` here could be `null`, so we should handle that case accordingly, otherwise the form won't work

**Review by timmyichen — CHANGES_REQUESTED** [review]

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx:81]: makes sense, I think by that logic none should be required?

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx:83]: ok, lmk what you think of the placeholders... didn't know what to put

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx:26]: should we re-route the user to another page or set up an error?

Comment (timmyichen) [reply pages/onboarding/index.tsx]: i would redirect to the home page, since we don't have a dedicated login page

File: pages/onboarding/index.tsx
Comment (timmyichen) [reply pages/onboarding/index.tsx:78]: placeholders should guide what the users should type in. i would make some (obviously) fake examples.  e.g. for linkedin we want the full URL so maybe something like `https://www.linkedin.com/in/your-profile`

File: pages/onboarding/index.tsx
Comment (timmyichen) [reply pages/onboarding/index.tsx]: yup - though we should ensure that at least _one_ field is entered before submitting.  i.e. the submit button should be disabled until they've typed in something for at least one field

Comment (jmoldyvan) [reply pages/onboarding/index.tsx]: is that for any specific reason?

Comment (timmyichen) [reply pages/onboarding/index.tsx]: just better UX - no point in allowing them to submit zero information

File: pages/onboarding/index.tsx
Comment (MainlyColors) [inline pages/onboarding/index.tsx:122]: The additional `Submit` here isn't needed , it causes `submitsubmit` to render

image

File: pages/onboarding/index.tsx
Comment (MainlyColors) [inline pages/onboarding/index.tsx:43]: The way this is implemented means that anytime the user changes the form the state is changed to `true` for setIsFormFilled even if the user removes the text they typed in. The button still becomes enabled so they can send another patch request with the same data they had before even if that data is exact same.

in this example, on the first `/onboarding` load after login, I filled out `bio, twitter username, linkedin url` and then came back to the page and filled out one of the empty fields partially then removed what I typed in. but I was still able to submit the data even though nothing changed
image

File: pages/onboarding/index.tsx
Comment (MainlyColors) [inline pages/onboarding/index.tsx:61]: if there was no user then the component would never get to this bit of code, it would fail at the initial `formData` state setup 
image

File: cypress/e2e/base.spec.ts
Comment (MainlyColors) [inline cypress/e2e/base.spec.ts:45]: I think you need to login first within the test to run this test correctly or intercept the request to add login data, I'm getting an error on the Cypress test since it visits `http://localhost:3000/onboarding` right away 

image

**Review by MainlyColors — COMMENTED** [review]
Just some things I found

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx:61]: got it, im hoping my new patch fixes it

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:35]: nit: `<boolean>` can be removed since it can be inferred from the initial value (`false`)

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:42]: if they type something in, then delete it, this will remain true.  `isFormFilled` can instead just be a map over all the values of the `formData` object:

```ts
const isFormFilled = Object.values(formData).every((value) => !!value);
```

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:51]: if there's an error for whatever reason, this code path won't get hit - will need to wrap the network call in a try/catch and `setIsLoading(false)` regardless of the outcome

Comment (timmyichen) [reply pages/onboarding/index.tsx:51]: Nevermind, I missed this!

**Comment by timmyichen** [comment]
@jmoldyvan how's this PR going? anything you need help with / blocked by anything?

File: pages/onboarding/index.tsx
Comment (devonzara) [inline pages/onboarding/index.tsx:73]: Can we move the form out of the header of the page? See `pages/profile/[id].tsx` for reference.

File: cypress/e2e/base.spec.ts
Comment (devonzara) [inline cypress/e2e/base.spec.ts:43]: This doesn't test that the submission works... it should `click()` the submit button, the redirect will be automatic if the form submission is a success.

Also, there's nothing testing the disabled/enabled state of the login button or the presence of the loading animation. Not sure if those are things we wish to test or if that's a bit too granular.

File: cypress/e2e/base.spec.ts
Comment (devonzara) [inline cypress/e2e/base.spec.ts:28]: This test will fail once we make `/onboarding` only accessible to logged-in users. Might want to add a `cy.login();` before the `visit()`.

That said, this isn't really testing much... do we even need it?

File: pages/onboarding/index.tsx
Comment (devonzara) [inline pages/onboarding/index.tsx:62]: We shouldn't really have `console` output on the client... if there may be an error here, we should probably show something on the screen for the user.

File: cypress/e2e/base.spec.ts
Comment (devonzara) [inline cypress/e2e/base.spec.ts:34]: This first `visit()` isn't necessary.

**Review by devonzara — CHANGES_REQUESTED** [review]
Few minor things... otherwise, looking good to me!

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx:73]: is there a benefit to doing this?

File: cypress/e2e/base.spec.ts
Comment (jmoldyvan) [reply cypress/e2e/base.spec.ts:28]: I mean, idk what the threshold for testing really is

File: pages/onboarding/index.tsx
Comment (devonzara) [inline pages/onboarding/index.tsx:62]: `alert()` doesn't create a very good user experience and should be avoided in most cases.

Since we're using MUI, a toast alert is pretty easy to do. It seems that MUI calls them [Snackbars](https://mui.com/material-ui/react-snackbar/).

File: cypress/e2e/onboarding.spec.ts
Comment (devonzara) [inline cypress/e2e/onboarding.spec.ts:16]: Too much was removed here, thus causing the tests to fail.

You had:
```js
    cy.visit('http://localhost:3000');
    cy.login();
    cy.visit('http://localhost:3000/onboarding');
    [...]
```
Only the first `cy.visit()` should have been removed, so:
```js
    cy.login();
    cy.visit('http://localhost:3000/onboarding');
    [...]
```

**Review by devonzara — CHANGES_REQUESTED** [review]

File: pages/onboarding/index.tsx
Comment (devonzara) [reply pages/onboarding/index.tsx:73]: A form like this shouldn't be in the header... making this change now will simplify the tasks later when we add proper page layouts to the application.

File: cypress/e2e/onboarding.spec.ts
Comment (timmyichen) [inline cypress/e2e/onboarding.spec.ts:12]: we can probably do without this test haha

File: cypress/e2e/onboarding.spec.ts
Comment (timmyichen) [inline cypress/e2e/onboarding.spec.ts:6]: is this necessary to do for onboarding related tests?

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:66]: i'm not sure if this logic is what we want - i'm reading this as: "submit is disabled if the form is not filled, and it is not loading" - meaning:
* if `isLoading` = true, then `isSubmitDisabled` = false

we can break this down into thinking: when do we want the button to be disabled?
* when the form is not filled (nothing to submit)
* when they're in the middle of a submission (avoid double submits)

these two criteria can be respectively expressed as:
* `!isFormFilled` (which you have), OR
* `isLoading`

```suggestion
  const disableSubmit = !isFormFilled || isLoading;
```

File: pages/onboarding/index.tsx
Comment (timmyichen) [reply pages/onboarding/index.tsx:73]: i also don't think we need a heading.  i.e. we don't need a title for the page that says "onboarding"... onboarding is mostly an internal concept.  we should be communicating this to our users in a more user-friendly way, like "Complete your profile to connect with other members!"

File: pages/onboarding/index.tsx
Comment (timmyichen) [inline pages/onboarding/index.tsx:106]: this technically isn't a valid username, also i think we should avoid gendered examples - would change this (plus the placeholder twitter handle above) to be more generic, e.g. `twitteruser` and `githubuser`

File: cypress/e2e/onboarding.spec.ts
Comment (devonzara) [reply cypress/e2e/onboarding.spec.ts:6]: I'd argue it doesn't hurt to clean up... get rid of any users created so there's a fresh slate for whatever tests may run after this file. Granted, it could be an `after()`...

File: pages/onboarding/index.tsx
Comment (devonzara) [inline pages/onboarding/index.tsx:71]: > Error: Objects are not valid as a React child (found: object with keys {message, name, code, config, request, response}). If you meant to render a collection of children, use an array instead.

`setOpenAlertError()` expects a `ReactChild`, not an object. A string would be a valid `ReactChild` so it should be `error.message`...

That said, we could probably just put a generic error in here... doing so would be less confusing to non-technical users.

File: pages/onboarding/index.tsx
Comment (devonzara) [inline pages/onboarding/index.tsx:150]: nit: I'd probably opt to have them in the bottom-right rather than the bottom-left. Personal opinion though...

**Review by devonzara — CHANGES_REQUESTED** [review]

File: pages/onboarding/index.tsx
Comment (MainlyColors) [reply pages/onboarding/index.tsx]: This is redirecting now but since it was done in a useEffect, the page loads completely and flashes at the user before redirecting.

File: pages/onboarding/index.tsx
Comment (MainlyColors) [reply pages/onboarding/index.tsx]: This still doesn't solve the issue because the state is queued so its always on a 1 character delay with this setup.

example, I have a empty form and write "test" in bio then backspace it

based on this code 
```ts
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setFormData(prevFormData => ({ ...prevFormData, [name]: value }));
    console.log(formData);
    const isFormFilled = Object.values(formData).some(value => !!value);
    console.log(isFormFilled);
    setIsFormFilled(isFormFilled);
  };
```
```
index.tsx:55 {bio: '', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 false
index.tsx:55 {bio: 't', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
index.tsx:55 {bio: 'te', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
index.tsx:55 {bio: 'tes', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
index.tsx:55 {bio: 'test', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
index.tsx:55 {bio: 'tes', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
index.tsx:55 {bio: 'te', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
index.tsx:55 {bio: 't', twitter_username: '', linkedin_url: '', github_username: '', website: ''}
index.tsx:57 true
```

at the end since there is a 1 character delay, I can still submit

image

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx:150]: yeah idk i liked it in the left. i guess we can consult design team?

Comment (devonzara) [reply pages/onboarding/index.tsx:150]: Fair, I believe my bias may come from operating systems as they tend to default to top-right or bottom-right.

Anyway, this is non-blocking and can always be changed later easily enough. 👍🏼

File: pages/onboarding/index.tsx
Comment (devonzara) [reply pages/onboarding/index.tsx:89]: @jmoldyvan Tim merged #95 which gives us proper layout components. When you merge changes from `main`, this file can be simplified.

**Review by devonzara — CHANGES_REQUESTED** [review]
Few changes suggested... Will also want to re-review after you merge `main` into this PR.

File: pages/onboarding/index.tsx
Comment (jmoldyvan) [reply pages/onboarding/index.tsx]: what do you mean? like we should be specializing our error catches instead of just in general firing on all server errors?

Comment (timmyichen) [reply pages/onboarding/index.tsx]: @devonzara not sure what you mean - validation errors in terms of validating the client input fields before sending the request off to the server?

Comment (devonzara) [reply pages/onboarding/index.tsx]: @jmoldyvan Sorta. Any input validation errors, specifically, should be handled separately.

@timmyichen Before is good too... but more importantly, when it hits the server. If input validation on the server fails, that should be reflected in the form so the user knows what to change and why.

Currently, if you remove the `maxlength="1000"` from the bio input field and type 1001 characters, you'll get this error - which is probably not an error we want to be sending back to the client.

image
```html
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Error<br> &nbsp; &nbsp;at Query.run (/home/user/workbench/dev-directory/node_modules/sequelize/src/dialects/postgres/query.js:76:25)<br> &nbsp; &nbsp;at /home/user/workbench/dev-directory/node_modules/sequelize/src/sequelize.js:650:28<br> &nbsp; &nbsp;at processTicksAndRejections (node:internal/process/task_queues:95:5)<br> &nbsp; &nbsp;at async PostgresQueryInterface.bulkUpdate (/home/user/workbench/dev-directory/node_modules/sequelize/src/dialects/abstract/query-interface.js:930:12)<br> &nbsp; &nbsp;at async Function.update (/home/user/workbench/dev-directory/node_modules/sequelize/src/model.js:3280:28)</pre>
</body>
</html>
```
In other words, I believe the toast errors should be used to show non-validation errors and validation errors should appear inline next to their respective fields.

Though, after seeing that there isn't really any server-side validation... I suppose this ought to be a separate ticket as that extends the scope of this quite a bit.

**Comment by timmyichen** [comment]
@jmoldyvan any updates?

**Review by timmyichen — APPROVED** [review]

**Review by Caleb-Cohen — APPROVED** [review]
LGTM!

**Pull request merged**
